### PR TITLE
pkg/policy: remove unused types from policy package

### PR DESCRIPTION
**Pull request by ianvernon**
Given that conntrack cleanup is no longer done upon policy updates from userspace, the L3L4Policy is no longer needed. Remove it for cleaner code, as well as `SecurityIDContexts`, `L4RuleContexts`, and `L7RuleContexts`. Refactor functions which use these types to remove references to them. This makes the code cleaner and easier to understand, which will help in refactoring the policy package in the future to separate the updating of the datapath from policy computation itself.

Signed-off by: Ian Vernon <ian@cilium.io>

### CODE REVIEW

**Comment by ianvernon** [comment]
test-me-please

**Comment by joestringer** [comment]
I thought that this is now a visibility thing - poorly named, but effectively a version of "realized state".

(I'm not necessarily arguing to keep it, just putting the thought out there)

**Comment by ianvernon** [comment]
@joestringer I'm not intending on merging this any time soon, this is more of an intellectual exercise for myself. Notice that I did not remove the code which generates what we have removed vs. what we have added during policy computation, as I believe this data can be quite useful. This is more of a "for-fun" PR :)

**Review by rlenglet — COMMENTED** [review]
I love it!

**Comment by rlenglet** [comment]
Are there any other uses of `SecurityIDContexts`? I think that type could now be removed altogether.

**Comment by rlenglet** [comment]
@joestringer How about we use the NPDS NetworkPolicy to represent the resolved policy for the endpoint instead? It currently contains the L4&L7 policy for the endpoint, and with https://github.com/cilium/cilium/pull/3634 it will also contain the L3 policy.

**Comment by ianvernon** [comment]
test-me-please

**Comment by ianvernon** [comment]
Build got stuck for some reason while building the docker image for K8s 1.7 tests: https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-Validated/2367/ . 
The code built just fine for runtime and K8s-1.10 tests, so am re-triggering a build .

**Comment by ianvernon** [comment]
test-me-please

File: pkg/endpoint/policy.go
Comment (raybejjani) [inline pkg/endpoint/policy.go:339]: The function comment isn't consistent anymore. Could you remove `and a map matching which rules were successfully added/modified;  and a map matching which rules were successfully removed.` ?

**Review by raybejjani — APPROVED** [review]
I'm probably not the best person to review, but I did give it a proper read. If anyone else wanted to review this, they should speak up. Otherwise, hopefully this unblocks an excellent PR.

**Comment by aanm** [comment]
@ianvernon does this means we can close https://github.com/cilium/cilium/issues/3393?

**Review by rlenglet — APPROVED** [review]
Awesome!

**Comment by tgraf** [comment]
> I thought that this is now a visibility thing - poorly named, but effectively a version of "realized state".

This is why I've left it in when removing the conntrack cleanup logic. I wasn't sure whether it would serve as help for whoever implements the desired vs realized state. At the very least it can serve as a starting point.

It looks like we are all agreeing on this aspect.

> How about we use the NPDS NetworkPolicy to represent the resolved policy for the endpoint instead? It currently contains the L4&L7 policy for the endpoint, and with #3634 it will also contain the L3 policy.

It still does not contain any CIDR logic. I would hold off to reuse it until we know that it will work as-is for CIDR. I understand and agree with @joestringer's plan and how it maps into NPDS but there is an open question regarding scale if somebody configures 60K of automatically generated `/32` CIDR rules.

**Comment by ianvernon** [comment]
test-me-please

**Comment by ianvernon** [comment]
test-me-please

**Comment by ianvernon** [comment]
test-me-please

**Review by joestringer — APPROVED** [review]

**Comment by ianvernon** [comment]
There was a small, trivial merge conflict involving documentation updates for `L3L4Policy` which was removed in the patch. Re-triggering tests, and will merge if tests pass, as there was no other rebasing needed.

**Comment by ianvernon** [comment]
test-me-please

**Pull request merged**
